File: src/Snapik.App/EditorModels.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Windows;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using System.Collections.Immutable;
using Snapik.Core.Models;
using CoreAnnotation = Snapik.Core.Models.AnnotationItem;
using CoreCapture = Snapik.Core.Models.CaptureItem;

namespace Snapik.App;

public enum EditorTool
{
    Select,
    Arrow,
    Rectangle,
    Pen,
    Highlight,
    Text,
    Conceal,
    Blur,
    Crop,
    Comment,
    // A mode of the panel and never the kind of a mark: the eraser removes what it is clicked on and
    // nothing carries this value into a session.
    Eraser
}

public sealed class AnnotationItem : INotifyPropertyChanged
{
    private string _note = string.Empty;
    private string _text = string.Empty;
    private bool _isSelected;

    public Guid Id { get; init; } = Guid.NewGuid();
    public EditorTool Kind { get; init; }
    public List<Point> Points { get; init; } = [];
    public List<List<Point>> AdditionalPathSegments { get; init; } = [];
    public Color Color { get; set; } = Color.FromRgb(49, 92, 245);
    public double Thickness { get; set; } = 4;
    public string Label { get; set; } = string.Empty;
    public Guid? ParentAnnotationId { get; set; }
    public string ArrowStyle { get; set; } = "straight";

    // The shift of the numbered badge from its automatic place, in image pixels; null is automatic.
    // A plain property on purpose: the drag of a note pill must not push a history entry per pixel.
    public Point? NoteOffset { get; set; }

    public AnnotationShape Shape { get; set; } = AnnotationShape.Rectangle;
    public AnnotationFill Fill { get; set; } = AnnotationFill.None;

    // The pattern of the stroke: solid, dashed or dotted. A mark without a stroke keeps the default
    // and writes it, so a highlighter or a caption reads back the same whatever was armed.
    public AnnotationLineStyle LineStyle { get; set; } = AnnotationLineStyle.Solid;

    // The colour inside the box; null means "the colour of the outline", which is how every mark
    // drawn before the fill had a colour of its own still reads.
    public Color? FillColor { get; set; }

    // A frame without an outline: a solid fill and no outline is what the conceal tool used to draw.
    public bool HasOutline { get; set; } = true;

    // The size a text mark is typed in, in the pixels of the capture.
    public double FontSize { get; set; } = TextMarkMetrics.DefaultFontSize;

    public string Note
    {
        get => _note;
        set { if (_note == value) return; _note = value; OnPropertyChanged(); }
    }

    public string Text
    {
        get => _text;
        set { if (_text == value) return; _text = value; OnPropertyChanged(); }
    }

    public bool IsSelected
    {
        get => _isSelected;
        set { if (_isSelected == value) return; _isSelected = value; OnPropertyChanged(); }
    }

    public AnnotationItem Clone() => new()
    {
        Id = Id,
        Kind = Kind,
        ParentAnnotationId = ParentAnnotationId, ArrowStyle = ArrowStyle, NoteOffset = NoteOffset,
        Shape = Shape, Fill = Fill, FillColor = FillColor, HasOutline = HasOutline, FontSize = FontSize,
        LineStyle = LineStyle,
        Points = [.. Points],
        AdditionalPathSegments = AdditionalPathSegments.Select(segment => segment.ToList()).ToList(),
        Color = Color,
        Thickness = Thickness,
        Label = Label,
        Note = Note,
        Text = Text,
        IsSelected = IsSelected
    };

    public CoreAnnotation ToCore(int imageWidth, int imageHeight) => new(
        Id,
        Kind switch
        {
            EditorTool.Comment => AnnotationKind.Comment,
            EditorTool.Arrow => AnnotationKind.Arrow,
            EditorTool.Rectangle => AnnotationKind.Rectangle,
            EditorTool.Pen => AnnotationKind.Freehand,
            EditorTool.Highlight => AnnotationKind.Highlight,
            EditorTool.Text => AnnotationKind.Text,
            EditorTool.Blur => AnnotationKind.Blur,
            _ => AnnotationKind.Rectangle
        },
        Points.Select(p => new NormalizedPoint(
            Math.Clamp(p.X / imageWidth, 0, 1),
            Math.Clamp(p.Y / imageHeight, 0, 1))).ToImmutableArray(),
        $"#{Color.A:X2}{Color.R:X2}{Color.G:X2}{Color.B:X2}",
        Thickness,
        Text,
        Note)
    {
        ParentAnnotationId = ParentAnnotationId, ArrowStyle = ArrowStyle,
        NoteOffset = NoteOffset is { } offset ? new NormalizedPoint(offset.X / imageWidth, offset.Y / imageHeight) : null,
        Shape = Shape, Fill = Fill, HasOutline = HasOutline, FontSize = FontSize, LineStyle = LineStyle,
        FillColor = FillColor is { } fillColor ? $"#{fillColor.A:X2}{fillColor.R:X2}{fillColor.G:X2}{fillColor.B:X2}" : null,
        PathSegments = AdditionalPathSegments.Count == 0 ? [] : new[] { Points }.Concat(AdditionalPathSegments)
            .Select(segment => segment.Select(p => new NormalizedPoint(Math.Clamp(p.X / imageWidth, 0, 1), Math.Clamp(p.Y / imageHeight, 0, 1))).ToImmutableArray())
            .ToImmutableArray()
    };

    public static AnnotationItem FromCore(CoreAnnotation item, int imageWidth, int imageHeight)
    {
        var segments = item.GetPathSegments()
            .Select(segment => segment.Select(p => new Point(p.X * imageWidth, p.Y * imageHeight)).ToList()).ToList();
        // A session written by a build that still had the conceal tool carries "redaction" marks.
        // The tool is gone; what it drew is a region with a solid black fill and no outline, and it
        // is written back in that shape the next time the session is saved.
        var redaction = item.Kind == AnnotationKind.Redaction;
        return new()
        {
        Id = item.Id,
        ParentAnnotationId = item.ParentAnnotationId, ArrowStyle = item.ArrowStyle,
        NoteOffset = item.NoteOffset is { } offset ? new Point(offset.X * imageWidth, offset.Y * imageHeight) : null,
        Shape = item.Shape,
        Fill = redaction ? AnnotationFill.Solid : item.Fill,
        FillColor = redaction ? Colors.Black : ParseFillColor(item.FillColor),
        HasOutline = !redaction && item.HasOutline,
        FontSize = item.FontSize,
        LineStyle = item.LineStyle,
        Kind = item.Kind switch
        {
            AnnotationKind.Comment => EditorTool.Comment,
            AnnotationKind.Arrow => EditorTool.Arrow,
            AnnotationKind.Rectangle => EditorTool.Rectangle,
            AnnotationKind.Freehand => EditorTool.Pen,
            AnnotationKind.Highlight => EditorTool.Highlight,
            AnnotationKind.Text => EditorTool.Text,
            AnnotationKind.Blur => EditorTool.Blur,
            _ => EditorTool.Rectangle
        },
        Points = segments.Count > 0 ? segments[0] : item.Points.Select(p => new Point(p.X * imageWidth, p.Y * imageHeight)).ToList(),
        AdditionalPathSegments = segments.Skip(1).ToList(),
        Color = (Color)ColorConverter.ConvertFromString(item.StrokeColor),
        Thickness = item.Thickness,
        Text = item.Text,
        Note = item.Note
        };
    }

    // A colour written by hand, or by a build that knew another format, means "the colour of the
    // outline" rather than a broken mark.
    private static Color? ParseFillColor(string? value)
    {
        if (string.IsNullOrWhiteSpace(value)) return null;
        try { return ColorConverter.ConvertFromString(value) is Color color ? color : null; }
        catch (Exception) { return null; }
    }

    public event PropertyChangedEventHandler? PropertyChanged;
    private void OnPropertyChanged([CallerMemberName] string? name = null) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

public sealed class CaptureItem : INotifyPropertyChanged
{
    private string _note = string.Empty;
    private bool _isSelected;
    private bool _isSent;
    private string _title = string.Empty;
    private CaptureKind _kind = CaptureKind.Region;
    private int _monitorCount;

    public Guid Id { get; init; } = Guid.NewGuid();
    public required BitmapSource Image { get; set; }
    public required string SourcePath { get; set; }
    public ObservableCollection<AnnotationItem> Annotations { get; } = [];
    public string DisplayLabel { get; set; } = "A";

    public string Note
    {
        get => _note;
        set { if (_note == value) return; _note = value; OnPropertyChanged(); }
    }

    public bool IsSelected
    {
        get => _isSelected;
        set { if (_isSelected == value) return; _isSelected = value; OnPropertyChanged(); }
    }

    // The capture was pasted with a package: it stays in the strip, dimmed and out of the next package.
    public bool IsSent
    {
        get => _isSent;
        set { if (_isSent == value) return; _isSent = value; OnPropertyChanged(); }
    }

    // Where the capture came from, how many monitors it covered and the name of the file it was
    // imported from: the chip of the card and the caption of the editor are bound to all three.
    public CaptureKind Kind
    {
        get => _kind;
        set { if (_kind == value) return; _kind = value; OnPropertyChanged(); }
    }

    public int MonitorCount
    {
        get => _monitorCount;
        set { var clamped = Math.Max(0, value); if (_monitorCount == clamped) return; _monitorCount = clamped; OnPropertyChanged(); }
    }

    public string Title
    {
        get => _title;
        set { if (_title == value) return; _title = value; OnPropertyChanged(); }
    }

    public int NoteCount => Annotations.Count(a => !string.IsNullOrWhiteSpace(a.Note)) + (string.IsNullOrWhiteSpace(Note) ? 0 : 1);

    public CaptureSnapshot Snapshot() => new(Id, Image, SourcePath, DisplayLabel, Note, Kind, MonitorCount, Title, Annotations.Select(a => a.Clone()).ToList());

    public CaptureItem DeepClone()
    {
        // The sent flag travels with the copy: a capture restored through "Undo" must not come back
        // as unsent and land in the next package a second time.
        var clone = new CaptureItem { Id = Id, Image = Image, SourcePath = SourcePath, DisplayLabel = DisplayLabel, Note = Note, IsSelected = IsSelected, IsSent = IsSent, Kind = Kind, MonitorCount = MonitorCount, Title = Title };
        foreach (var annotation in Annotations.Select(a => a.Clone())) clone.Annotations.Add(annotation);
        return clone;
    }

    public CoreCapture ToCore() => new(
        Id,
        SourcePath,
        Image.PixelWidth,
        Image.PixelHeight,
        Image.DpiX > 0 ? Image.DpiX : 96,
        Image.DpiY > 0 ? Image.DpiY : 96,
        Title,
        Note,
        Annotations.Select(a => a.ToCore(Image.PixelWidth, Image.PixelHeight)).ToImmutableArray())
    {
        Sent = IsSent,
        Kind = Kind,
        MonitorCount = MonitorCount
    };

    public static CaptureItem FromCore(CoreCapture item, BitmapSource image)
    {
        var capture = new CaptureItem
        {
            Id = item.Id, SourcePath = item.SourceImagePath, Image = image, Note = item.Note, IsSent = item.Sent,
            Kind = item.Kind, MonitorCount = item.MonitorCount, Title = item.Title
        };
        foreach (var annotation in item.Annotations)
            capture.Annotations.Add(AnnotationItem.FromCore(annotation, image.PixelWidth, image.PixelHeight));
        return capture;
    }

    public void Restore(CaptureSnapshot snapshot)
    {
        Image = snapshot.Image;
        SourcePath = snapshot.SourcePath;
        DisplayLabel = snapshot.DisplayLabel;
        Note = snapshot.Note;
        Kind = snapshot.Kind;
        MonitorCount = snapshot.MonitorCount;
        Title = snapshot.Title;
        Annotations.Clear();
        foreach (var annotation in snapshot.Annotations.Select(a => a.Clone())) Annotations.Add(annotation);
    }

    public event PropertyChangedEventHandler? PropertyChanged;
    private void OnPropertyChanged([CallerMemberName] string? name = null) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

public sealed record CaptureSnapshot(Guid CaptureId, BitmapSource Image, string SourcePath, string DisplayLabel, string Note, CaptureKind Kind, int MonitorCount, string Title, IReadOnlyList<AnnotationItem> Annotations);

public sealed record PreparedPackage(Guid ExportId, IReadOnlyList<string> ImagePaths, string PromptText, string DirectoryPath);

File: macos/Sources/SnapikMac/Editor/EditorModels.swift
// Port of `src/Snapik.App/EditorModels.cs`, SPEC §2.4, §1.3 (tool table), §1.5 (history unit)
import AppKit
import SnapikCore

/// Port of `EditorTool` (`EditorModels.cs:17-28`). Raw values are the single-letter hotkeys used
/// by `AnnotationCanvasView`/`OverlayEditorController+Keys` (SPEC §7.5/§7.6): `V R A B C P H T X`.
enum EditorTool: String {
    case select = "V"
    case rectangle = "R"
    case arrow = "A"
    case pen = "P"
    case highlight = "H"
    case text = "T"
    case conceal = "X"
    case blur = "B"
    case crop = "C"
    /// Port of `EditorTool.Comment` (SPEC-DELTA-2.md §1.3, SPEC-DELTA-2B.md §C2). One-shot: a
    /// single click places a pin and returns to `.select` (`OverlayEditorController+Chips.swift`).
    case comment = "N"
    /// Port of `EditorTool.Eraser` (`EditorModels.cs:29`, SPEC-DELTA-3 §1.4 E-5): a mode of the
    /// panel and never the kind of a mark. The eraser removes what it is clicked on and nothing
    /// carries this value into a session.
    case eraser = "E"

    /// Port of `IsDrawingTool` (`AnnotationCanvas.cs:355`): everything but Select and the eraser
    /// puts something on the capture, and shows the crosshair over it.
    var isDrawing: Bool { self != .select && self != .eraser }
}

/// Port of `AnnotationItem` (`EditorModels.cs:30-131`). A plain reference type (no
/// `INotifyPropertyChanged` binding machinery — the AppKit views that own one of these call
/// `needsDisplay = true` / the relevant reposition helpers directly after mutating it).
///
/// Points are stored in **image pixel space** (not normalized), matching the Windows `Point`
/// convention; conversion to/from Core's normalized `[0,1]` space happens only in `toCore`/
/// `fromCore` (SPEC §2.4).
final class EditorAnnotation {
    let id: SBGuid
    var kind: EditorTool
    var points: [CGPoint]
    var additionalPathSegments: [[CGPoint]]
    var color: NSColor
    var thickness: Double
    var label: String = ""
    var note: String = ""
    var text: String
    /// Port of `AnnotationItem.ParentAnnotationId` (SPEC-DELTA-2.md §2.1). `nil` = comment attached
    /// to the whole capture ("к снимку"); non-nil = attached to that annotation ("к отметке").
    var parentAnnotationId: SBGuid?
    /// Port of `AnnotationItem.ArrowStyle` (SPEC-DELTA-2.md §2.1): `"straight"`/`"curved"`/
    /// `"bold"`/`"wide"`. Meaningless for non-arrow kinds, always carried along regardless.
    var arrowStyle: String
    /// Port of `AnnotationItem.NoteOffset` (`EditorModels.cs:50`): the shift of the numbered badge
    /// from its automatic place, **in image pixels**; `nil` is automatic. A plain property on
    /// purpose: the drag of a note pill must not push a history entry per pixel.
    var noteOffset: CGPoint?
    /// Port of `Shape`/`Fill`/`FillColor`/`LineStyle`/`FontSize` (`EditorModels.cs:52-68`).
    var shape: AnnotationShape
    var fill: AnnotationFill
    /// The colour inside the box; `nil` means "the colour of the outline", which is how every mark
    /// drawn before the fill had a colour of its own still reads.
    var fillColor: NSColor?
    var lineStyle: AnnotationLineStyle
    /// The size a caption is typed in, in the pixels of the capture.
    var fontSize: Double

    init(
        id: SBGuid = SBGuid(),
        kind: EditorTool,
        points: [CGPoint],
        additionalPathSegments: [[CGPoint]] = [],
        color: NSColor,
        thickness: Double,
        text: String? = nil,
        note: String = "",
        parentAnnotationId: SBGuid? = nil,
        arrowStyle: String = "straight",
        noteOffset: CGPoint? = nil,
        shape: AnnotationShape = .rectangle,
        fill: AnnotationFill = .none,
        fillColor: NSColor? = nil,
        lineStyle: AnnotationLineStyle = .solid,
        fontSize: Double = TextMarkMetrics.defaultFontSize
    ) {
        self.id = id
        self.kind = kind
        self.points = points
        self.additionalPathSegments = additionalPathSegments
        self.color = color
        self.thickness = thickness
        self.text = text ?? ""
        self.note = note
        self.parentAnnotationId = parentAnnotationId
        self.arrowStyle = arrowStyle
        self.noteOffset = noteOffset
        self.shape = shape
        self.fill = fill
        self.fillColor = fillColor
        self.lineStyle = lineStyle
        self.fontSize = fontSize
    }

    /// Port of `AnnotationItem.Clone()` (`:62-74`). Identity (`id`) is preserved, matching the
    /// Windows deep-clone used for undo/redo snapshots.
    func clone() -> EditorAnnotation {
        let copy = EditorAnnotation(
            id: id,
            kind: kind,
            points: points,
            additionalPathSegments: additionalPathSegments,
            color: color,
            thickness: thickness,
            text: text,
            note: note,
            parentAnnotationId: parentAnnotationId,
            arrowStyle: arrowStyle,
            noteOffset: noteOffset,
            shape: shape,
            fill: fill,
            fillColor: fillColor,
            lineStyle: lineStyle,
            fontSize: fontSize)
        // The letter in the badge is state of the view and not of the model, but an undo that gave
        // it back blank would redraw every badge as a placeholder until the next `refreshLabels`.
        copy.label = label
        return copy
    }

    /// Port of `EditorTool` -> `AnnotationKind` (SPEC §2.4 table; `EditorModels.cs:78-88`).
    /// Select/Crop fall back to `.rectangle`, matching the C# `_ => AnnotationKind.Rectangle`.
    var coreKind: AnnotationKind { Self.coreKind(of: kind) }

    /// The same mapping without a mark to read it from: the panel asks it of the tool in the hand.
    static func coreKind(of tool: EditorTool) -> AnnotationKind {
        switch tool {
        case .arrow: return .arrow
        case .rectangle: return .rectangle
        case .pen: return .freehand
        case .highlight: return .highlight
        case .text: return .text
        case .conceal: return .redaction
        case .blur: return .blur
        case .comment: return .comment
        case .select, .crop, .eraser: return .rectangle
        }
    }

    /// Port of `ToCore` (`:76-100`, SPEC §2.4): points divided by image size and clamped to
    /// `[0,1]`; color formatted `#AARRGGBB` uppercase; `pathSegments` only populated when there
    /// are additional segments, with the primary `points` set included first.
    func toCore(imageWidth: Int, imageHeight: Int) -> AnnotationItem {
        let width = max(1, imageWidth)
        let height = max(1, imageHeight)
        func normalize(_ p: CGPoint) -> NormalizedPoint {
            NormalizedPoint(
                min(max(Double(p.x) / Double(width), 0), 1),
                min(max(Double(p.y) / Double(height), 0), 1))
        }

        let normalizedPoints = points.map(normalize)
        var pathSegments: [[NormalizedPoint]] = []
        if !additionalPathSegments.isEmpty {
            pathSegments.append(normalizedPoints)
            pathSegments.append(contentsOf: additionalPathSegments.map { $0.map(normalize) })
        }

        return AnnotationItem(
            id: id,
            kind: coreKind,
            points: normalizedPoints,
            strokeColor: color.hexARGB,
            thickness: thickness,
            text: text,
            note: note,
            pathSegments: pathSegments,
            parentAnnotationId: parentAnnotationId,
            arrowStyle: arrowStyle,
            // A shift, not a coordinate: it is divided by the size of the image and never clamped
            // into `[0,1]` (SPEC-DELTA-3 §2.1).
            noteOffset: noteOffset.map { NormalizedPoint(Double($0.x) / Double(width), Double($0.y) / Double(height)) },
            shape: shape,
            fill: fill,
            lineStyle: lineStyle,
            fillColor: fillColor?.hexARGB,
            // [ТЗ№4 D1] The switch is gone from the panel and the flag with it: `hasOutline` is
            // read and never written any more (SPEC-DELTA-4 §2.2).
            fontSize: fontSize)
    }

    /// Port of `AnnotationItem.FromCore` (`:102-127`).
    static func fromCore(_ item: AnnotationItem, imageWidth: Int, imageHeight: Int) -> EditorAnnotation {
        let width = Double(imageWidth)
        let height = Double(imageHeight)
        func denormalize(_ p: NormalizedPoint) -> CGPoint { CGPoint(x: p.x * width, y: p.y * height) }

        let segments = item.getPathSegments().map { segment in segment.map(denormalize) }
        let points = segments.first ?? item.points.map(denormalize)
        let additional = segments.count > 1 ? Array(segments.dropFirst()) : []

        let kind: EditorTool
        switch item.kind {
        case .arrow: kind = .arrow
        case .rectangle: kind = .rectangle
        case .freehand: kind = .pen
        case .highlight: kind = .highlight
        case .text: kind = .text
        // A session written by a build that still had the conceal tool carries "redaction" marks.
        // The tool is gone; what it drew is a region with a solid black fill, and it is written back
        // in that shape the next time the session is saved (SPEC-DELTA-3 §2.1).
        case .redaction: kind = .rectangle
        case .blur: kind = .blur
        case .comment: kind = .comment
        }
        let redaction = item.kind == .redaction
        // [ТЗ№4 D1] `hasOutline: false` on a boxed mark reads as "a solid fill of one colour"; the
        // field is never written again (SPEC-DELTA-4 §2.2). The colour of that fill is its own if
        // the file carries one, the colour of the stroke otherwise.
        let solidWithoutOutline = item.kind == .rectangle && item.legacyHasOutline == false
        let strokeColor = NSColor(argbHex: item.strokeColor)

        return EditorAnnotation(
            id: item.id,
            kind: kind,
            points: points,
            additionalPathSegments: additional,
            color: strokeColor,
            thickness: item.thickness,
            text: item.text,
            note: item.note,
            parentAnnotationId: item.parentAnnotationId,
            arrowStyle: item.arrowStyle,
            noteOffset: item.noteOffset.map { CGPoint(x: $0.x * width, y: $0.y * height) },
            shape: item.shape,
            fill: redaction || solidWithoutOutline ? .solid : item.fill,
            fillColor: redaction
                ? NSColor(srgbRed: 0, green: 0, blue: 0, alpha: 1)
                : (EditorAppearance.parseFillColor(item.fillColor) ?? (solidWithoutOutline ? strokeColor : nil)),
            lineStyle: item.lineStyle,
            fontSize: item.fontSize)
    }
}

/// Port of `CaptureItem` (`EditorModels.cs:133-198`).
final class EditorCapture {
    let id: SBGuid
    var image: CGImage
    var sourceImagePath: String
    var displayLabel: String = "A"
    /// Port of `CaptureItem.Title` (`EditorModels.cs:133-198`). The name of the file an imported
    /// capture came from, which the caption of the editor says out loud (SPEC-DELTA-4 §3.5); it is
    /// threaded through `toCore`/`fromCore` so a reopened capture's title round-trips instead of
    /// being silently blanked on every commit (finding R7).
    var title: String = ""
    /// Port of `CaptureItem.Kind`/`MonitorCount` (SPEC-DELTA-4 §2.1, E-7): what the capture is, and
    /// how many monitors a whole-screen one covered. The editor reads both for its caption and
    /// carries them back untouched — the strip is what sets them.
    var kind: CaptureKind = .region
    var monitorCount = 0
    var note: String = ""
    var annotations: [EditorAnnotation] = []
    var dpiX: Double
    var dpiY: Double

    init(id: SBGuid = SBGuid(), image: CGImage, sourceImagePath: String, dpiX: Double = 96, dpiY: Double = 96) {
        self.id = id
        self.image = image
        self.sourceImagePath = sourceImagePath
        self.dpiX = dpiX
        self.dpiY = dpiY
    }

    /// Port of `NoteCount` (`:156`): annotations with a non-blank note, plus one if the capture's
    /// own note is non-blank.
    var noteCount: Int {
        annotations.filter { !$0.note.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }.count
            + (note.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? 0 : 1)
    }

    /// Port of `CaptureItem.Snapshot()` (`:158`) — used as the undo/redo history unit (SPEC §1.5).
    func snapshot() -> CaptureSnapshot {
        CaptureSnapshot(
            captureId: id,
            image: image,
            sourceImagePath: sourceImagePath,
            displayLabel: displayLabel,
            note: note,
            annotations: annotations.map { $0.clone() })
    }

    /// Port of `CaptureItem.Restore(CaptureSnapshot)` (`:186-194`).
    func restore(_ snapshot: CaptureSnapshot) {
        image = snapshot.image
        sourceImagePath = snapshot.sourceImagePath
        displayLabel = snapshot.displayLabel
        note = snapshot.note
        annotations = snapshot.annotations.map { $0.clone() }
    }

    /// Port of `CaptureItem.ToCore()` (`:167-176`). `title` round-trips whatever `fromCore` last
    /// set (finding R7) rather than always writing `""`.
    func toCore() -> CaptureItem {
        CaptureItem(
            id: id,
            sourceImagePath: sourceImagePath,
            pixelWidth: image.width,
            pixelHeight: image.height,
            dpiX: dpiX > 0 ? dpiX : 96,
            dpiY: dpiY > 0 ? dpiY : 96,
            title: title,
            note: note,
            annotations: annotations.map { $0.toCore(imageWidth: image.width, imageHeight: image.height) },
            kind: kind,
            monitorCount: monitorCount)
    }

    /// Port of `CaptureItem.FromCore(CoreCapture, BitmapSource)` (`:178-184`).
    static func fromCore(_ item: CaptureItem, image: CGImage) -> EditorCapture {
        let capture = EditorCapture(id: item.id, image: image, sourceImagePath: item.sourceImagePath, dpiX: item.dpiX, dpiY: item.dpiY)
        capture.title = item.title
        capture.kind = item.kind
        capture.monitorCount = item.monitorCount
        capture.note = item.note
        capture.annotations = item.annotations.map { EditorAnnotation.fromCore($0, imageWidth: image.width, imageHeight: image.height) }
        return capture
    }
}

/// Port of `CaptureSnapshot` (`EditorModels.cs:200`) — the undo/redo unit for the whole capture.
struct CaptureSnapshot {
    let captureId: SBGuid
    let image: CGImage
    let sourceImagePath: String
    let displayLabel: String
    let note: String
    let annotations: [EditorAnnotation]
}

/// Port of `OverlaySnapshot` (`OverlayEditorWindow.xaml.cs:794`) — SPEC §1.5: the full local
/// undo/redo unit, one level above `CaptureSnapshot` (also captures the crop rect and which chips
/// were open).
struct OverlaySnapshot {
    let capture: CaptureSnapshot
    /// Crop rect in the active editing window's local point space (top-left origin, Y down —
    /// see `EditorGeometry`).
    let cropRect: CGRect
    let visibleChipIds: Set<SBGuid>
}

// MARK: - Color <-> `#AARRGGBB` hex (SPEC §2.4)

extension NSColor {
    /// Port of `$"#{Color.A:X2}{Color.R:X2}{Color.G:X2}{Color.B:X2}"`.
    var hexARGB: String {
        guard let rgb = usingColorSpace(.deviceRGB) else { return "#FF000000" }
        let a = UInt8((rgb.alphaComponent * 255).rounded())
        let r = UInt8((rgb.redComponent * 255).rounded())
        let g = UInt8((rgb.greenComponent * 255).rounded())
        let b = UInt8((rgb.blueComponent * 255).rounded())
        return String(format: "#%02X%02X%02X%02X", a, r, g, b)
    }

    /// Port of `$"#{Color.R:X2}{Color.G:X2}{Color.B:X2}"` (`OverlayEditorWindow.Appearance.cs:55`)
    /// — the appearance popover's hex field display format: RGB only, no alpha, unlike
    /// `hexARGB` (the annotation-storage format).
    var hexRGB: String {
        guard let rgb = usingColorSpace(.deviceRGB) else { return "#000000" }
        let r = UInt8((rgb.redComponent * 255).rounded())
        let g = UInt8((rgb.greenComponent * 255).rounded())
        let b = UInt8((rgb.blueComponent * 255).rounded())
        return String(format: "#%02X%02X%02X", r, g, b)
    }

    /// Port of `(Color)ColorConverter.ConvertFromString(item.StrokeColor)` for the `#AARRGGBB`
    /// format this app always writes. Falls back to opaque black on malformed input.
    convenience init(argbHex: String) {
        var value: UInt64 = 0
        let hexString = argbHex.hasPrefix("#") ? String(argbHex.dropFirst()) : argbHex
        Scanner(string: hexString).scanHexInt64(&value)
        guard hexString.count == 8 else {
            self.init(srgbRed: 0, green: 0, blue: 0, alpha: 1)
            return
        }
        let a = CGFloat((value & 0xFF00_0000) >> 24) / 255.0
        let r = CGFloat((value & 0x00FF_0000) >> 16) / 255.0
        let g = CGFloat((value & 0x0000_FF00) >> 8) / 255.0
        let b = CGFloat(value & 0x0000_00FF) / 255.0
        self.init(srgbRed: r, green: g, blue: b, alpha: a)
    }
}
